### DISCUSSION TITLE
Broken navigation after upgrade to 10.3

### DISCUSSION THREAD

**Discussion by simdoor** [pimcore/pimcore#12561] [Q&A / Support]
Hi,
I came across [this pull request ](https://github.com/pimcore/pimcore/pull/11149)while debugging an issue on a project.
I see the point of the change, but in my case the navigation tree does not mirror the URI structure.

Example:

|- Product page /our-offer
          |
          |- Product 1 /products/product-1
          |- Product 2 /products/product-2

When I visit page _/products/product-1_ it is not marked as active since the the PrefixRecursiveFilterIterator skips the Product page. I fixed it by extending the Builder class and overriding the findActivePages, just wanted to point it out in case anyonehas similar issue

**Comment by jdreesen** [comment]
Can you please describe your setup a bit more, and why it doesn't match with the URI structure? Maybe there's something we can do about it.

**Comment by simdoor** [comment]
Hi @jdreesen
sorry for late reply. Unfortunately I cannot simply copy&paste the project code, but I'll try to explain the setup.

I build my navigation tree using the callback function to inject objects pretty much as explained in [Pimcore documentation](https://pimcore.com/docs/pimcore/current/Development_Documentation/Documents/Navigation.html#page_Adding-Custom-Items-to-the-Navigation):


        
```
          /**
           * @param Document $document
           * @param Document $startNode
           *
           * @return \Pimcore\Navigation\Container
           * @throws \Exception
           */
          public function getNewsLinks(Document $document, Document $startNode)
          {
              $navigation = $this->navigationHelper->build([
                  'active' => $document,
                  'root' => $startNode,
                  'pageCallback' => function($page, $document) {
                      /** @var \Pimcore\Model\Document $document */
                      /** @var \Pimcore\Navigation\Page\Document $page */
                      if($document->getProperty("templateType") == "news") {
                          $list = new \Pimcore\Model\DataObject\News\Listing;
                          $list->load();
                          foreach($list as $news) {
                              $detailLink = $this->newsLinkGenerator->generate($news, ['document' => $document]);
                              $uri = new \Pimcore\Navigation\Page\Document([
                                  "label" => $news->getTitle(),
                                  "id" => "object-" . $news->getId(),
                                  "uri" => $detailLink,
                              ]);
                              $page->addPage($uri);
                          }
                      }
                  }
              ]);
      
              return $navigation;
          }
```

Since my document tree looks something like this:
<img width="318" alt="document-tree" src="https://user-images.githubusercontent.com/7021742/178989405-039d6c65-d7d3-4718-9d27-2c7c6f71bf20.png">

This will be the resulting html when I visit _mydomain.com/who-we-are/news/my-first-news_. You will notice that both the _active_ and the  _active-trail_ classes are at the right place
 
```
<ul class="navbar-level-1">
    <li>
        <a class="navbar-item active active-trail" href="/who-we-are" >Who we are </a>
        <ul class="navbar-multi-column">
            <li>
                <a class="navbar-item navbar-item--level-2 active active-trail" href="/who-we-are/news">News</a>

                <ul class="navbar-multi-column">
                    <li>
                        <a class="navbar-item navbar-item--level-3 active" href="/who-we-are/news/my-first-news" >My First News</a>
                    </li>
                </ul>

            </li>
        </ul>
    </li>
</ul>
``` 

However in my context the NewsLinkGenerator is not generating a url relative to the News document, but a kind of static URL, something like  _/our-news/my-first-news_

As a result, when I visit  _mydomain.com/our-news/my-first-news_ the findActivePages method that I originally mentioned, fails to recognise the URL as active, due to the PrefixRecursiveFilterIterator. When I loop through navigation items to output the navigation, the _active_ classes are missing


```
<ul class="navbar-level-1">
    <li>
        <a class="navbar-item" href="/who-we-are" >Who we are </a>
        <ul class="navbar-multi-column">
            <li>
                <a class="navbar-item navbar-item--level-2" href="/who-we-are/news">News</a>

                <ul class="navbar-multi-column">
                    <li>
                        <a class="navbar-item navbar-item--level-3" href="/our-news/my-first-news" >My First News</a>
                    </li>
                </ul>

            </li>
        </ul>
    </li>
</ul>
```


I understand this could be considered an edge case, that might be better addressed by generating a relative link for the News object.

Hope this makes more sense, let me know if otherwise.

**Reply by jdreesen** [reply]
It makes sense, yes, thx for the explanation! I hadn't thought of this use case yet.

We could probably find a solution for this. The "easiest" way may be to skip the `PrefixRecursiveFilterIterator` if pages are added via the `pageCallback`, because we can't be sure that it's Pimcore's tree structure anymore, which we can filter this way.

If @brusch, @dvesh3 or anyone else from the Pimcore team think it's worth it (and not too edge-casey) I'd try to find a solution to fix this. But it won't be too soon, as I'm rather busy at the moment (and on vacation in August).
